test/controllers/pensions/UpdatePensionProviderControllerSpec.scala

```diff
@@ -425,12 +363,136 @@ class UpdatePensionProviderControllerSpec extends PlaySpec with FakeTaiPlayAppli
+          thenReturn(Future.successful(TaiSuccessResponseWithPayload(Seq(pensionTaxCodeIncome, empTaxCodeIncome))))
+
+        val result = sut.redirectUpdatePension(4)(RequestBuilder.buildFakeRequestWithAuth("GET"))
+
+        status(result) mustBe INTERNAL_SERVER_ERROR
+      }
+    }
+  }
+
+  "duplicateSubmissionWarning" must {
+    "show duplicateSubmissionWarning view" in {
+      val updatePensionTest = createSUT
+      val pensionId = 1
+
+      when(journeyCacheService.mandatoryValues(Matchers.anyVararg[String])(any()))
+        .thenReturn(Future.successful(Seq(pensionName, pensionId.toString)))
+
+      val result = updatePensionTest.duplicateSubmissionWarning(RequestBuilder.buildFakeRequestWithAuth("GET"))
+      val doc = Jsoup.parse(contentAsString(result))
+
+      status(result) mustBe OK
+      doc.title() must include(Messages("tai.pension.warning.customGaTitle"))
+    }
+  }
+
+  "submitDuplicateSubmissionWarning" must {
+    "redirect to the update remove employment decision page" when {
+      "I want to update my employment is selected" in {
+        val pensionId = 1
+        val updatePensionTest = createSUT
+
+        when(journeyCacheService.mandatoryValues(Matchers.anyVararg[String])(any()))
+          .thenReturn(Future.successful(Seq(pensionName, pensionId.toString)))
+
+        val result = updatePensionTest.submitDuplicateSubmissionWarning(RequestBuilder.buildFakeRequestWithAuth("POST")
+          .withFormUrlEncodedBody(YesNoChoice -> YesValue))
+
+        status(result) mustBe SEE_OTHER
+        redirectLocation(result).get mustBe controllers.pensions.routes.UpdatePensionProviderController.doYouGetThisPension().url
+      }
+    }
+
+    "redirect to the income source summary page" when {
+      "I want to return to my employment details is selected" in {
+        val pensionId = 1
+        val updatePensionTest = createSUT
+
+        when(journeyCacheService.mandatoryValues(Matchers.anyVararg[String])(any()))
+          .thenReturn(Future.successful(Seq(pensionName, pensionId.toString)))
+
+        val result = updatePensionTest.submitDuplicateSubmissionWarning(RequestBuilder.buildFakeRequestWithAuth("POST")
+          .withFormUrlEncodedBody(YesNoChoice -> NoValue))
+
+        status(result) mustBe SEE_OTHER
+        redirectLocation(result).get mustBe controllers.routes.IncomeSourceSummaryController.onPageLoad(pensionId).url
+      }
+    }
+
+    "return BadRequest" when {
+      "there is a form validation error (standard form validation)" in {
+        val pensionId = 1
+        val updatePensionTest = createSUT
+
+        when(journeyCacheService.mandatoryValues(Matchers.anyVararg[String])(any()))
+          .thenReturn(Future.successful(Seq(pensionName, pensionId.toString)))
+
+        val result = updatePensionTest.submitDuplicateSubmissionWarning(RequestBuilder.buildFakeRequestWithAuth("POST")
+          .withFormUrlEncodedBody(YesNoChoice -> ""))
```

Comment:
Dedupe test set up

Reply:
Deduped the fake request, journeyCache Call and taxAccountService Call now.